mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala

```diff
@@ -213,32 +221,36 @@ class StringIndexer @Since("1.4.0") (
     val labelsArray = $(stringOrderType) match {
       case StringIndexer.frequencyDesc =>
         val sortFunc = StringIndexer.getSortFunc(ascending = false)
-        countByValue(dataset, inputCols).map { counts =>
+        val orgStrings = countByValue(dataset, inputCols).toSeq
+        ThreadUtils.parmap(orgStrings, "sortingStringLabels", 8) { counts =>
           counts.toSeq.sortWith(sortFunc).map(_._1).toArray
-        }
+        }.toArray
       case StringIndexer.frequencyAsc =>
```

Comment:
PS I wonder if we can reduce the code duplication while we're at it. Each pair of cases is almost identical. Maybe they can share the handling and just vary one element (asc/desc) by checking the actual value one more time in the block

Reply:
Done. Dedup the code now.